### PR TITLE
Challenge/count_string_letters

### LINKED ISSUE #18: Count Specific Letters in a String
**Program Title:**
Letter Count

**Description:**
Write a program that counts how many times a specific letter (provided by the user) appears in a string. 

**Expected Outcome:**
Returns a string that says "The letter '{letter}' appears {count} times." 

**Helpful Resources:**
https://www.w3schools.com/python/ref_string_count.asp

### PR DESCRIPTION
**Pull request by Hopeeid**
---
name:  Count string letter
about: A function that counts specific letters in a string.
---

Resolves #18 

## Behavior
-This function takes a sentence and a letter and counts the number of times the letter appears in the sentence.
- The input can only be a string otherwise the function would result in an error
- The output is a string containing the number of times the letter appears in the given sentence.

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/count_specific_string_letters.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        AssertionError: if the argument(sentence) is not a string
+        ValueError: If the letters parameter is an empty string.
+
+    Example 1:
+    >>> count_string("hello", "l")
+    'the letter 'l' appears 2 times'
+
+    Example 2:
+    >>> count_string("I love pineapples too much", "t"))
+    'the letter 't' appears 1 times'
```
Comment (ndubuisia) [inline solutions/count_specific_string_letters.py]: The docstring should include at least 3 doctests.

File: solutions/count_specific_string_letters.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+    """
+    if not letters:
+        raise ValueError("The letter to count must not be empty.")
```
Comment (ndubuisia) [inline solutions/count_specific_string_letters.py]: In the function ValueError is raised for wrong value input but AssertionError and ValueError were documented.

File: solutions/tests/test_count_specific_string_letters.py
Comment (ndubuisia) [inline solutions/tests/test_count_specific_string_letters.py:1]: A nice test you have got here @Hopeeid. Just wanted to know why you included an example in the docstring of each test.

**Review by ndubuisia — COMMENTED** [review]
A great code you have got here @Hopeeid. I wanted to let you know that there is no module header in each file.
Also, I wanted to ask why you chose to return a sentence when you could have just returned an integer.

File: solutions/tests/test_count_specific_string_letters.py
Comment (Hopeeid) [reply solutions/tests/test_count_specific_string_letters.py:1]: Thank you for your review. I included an example in the docstring of each test just for extra details. Or is it not a good practice?

File: solutions/count_specific_string_letters.py (hunk 2 above)
Comment (Hopeeid) [reply solutions/count_specific_string_letters.py]: Thank you for your review. It was an oversight and it has been corrected.

File: solutions/count_specific_string_letters.py (hunk 1 above)
Comment (Hopeeid) [reply solutions/count_specific_string_letters.py]: Thank you for your review. I've corrected this by adding a third doctest.

Comment (danielluyi) [reply solutions/count_specific_string_letters.py]: Welldone Hope. I just confirmed this as well

**Review by danielluyi — COMMENTED** [review]
Welldone Hope. I observed you've made adjustments based on orevious recommendations

**Review by danielluyi — APPROVED** [review]
Welldone Hope
